notifications_utils/logging.py

```diff
@@ -17,11 +19,63 @@
 logger = logging.getLogger(__name__)
 
 
+def _common_request_extra_log_context():
+    return {
+        "method": request.method,
+        "url": request.url,
+        "endpoint": request.endpoint,
+        # pid and is available on LogRecord by default, as `process` but I don't see
+        # a straightforward way of selectively including it only in certain log messages -
+        # it is designed to be included when the formatter is being configured. This is
+        # why I'm manually grabbing it and putting it in as `extra` here, avoiding the
+        # existing parameter name to prevent LogRecord from complaining
+        "process_": getpid(),
+    }
+
+
 def init_app(app, statsd_client=None, extra_filters: Sequence[logging.Filter] = tuple()):
     app.config.setdefault("NOTIFY_LOG_LEVEL", "INFO")
     app.config.setdefault("NOTIFY_APP_NAME", "none")
     app.config.setdefault("NOTIFY_LOG_PATH", "./log/application.log")
     app.config.setdefault("NOTIFY_RUNTIME_PLATFORM", None)
+    app.config.setdefault(
+        "NOTIFY_REQUEST_LOG_LEVEL",  # noqa
+        "WARNING" if app.config["NOTIFY_RUNTIME_PLATFORM"] == "paas" else "NOTSET",
+    )
+
+    @app.before_request
+    def before_request():
+        # annotating this onto request instead of flask.g as it probably shouldn't
+        # be inheritable from a request-less application context
+        request.before_request_real_time = time.perf_counter()
+
+        # emit an early log message to record that the request was received by the app
+        context = _common_request_extra_log_context()
+        current_app.logger.getChild("request").log(
+            logging.DEBUG,
+            "Received request %(method)s %(url)s",
+            context,
+            extra=context,
+        )
+
+    @app.after_request
+    def after_request(response):
+        context = {
+            "status": response.status_code,
+            "duration_real": (
+                (time.perf_counter() - request.before_request_real_time)
+                if hasattr(request, "before_request_real_time")
+                else None
+            ),
+            **_common_request_extra_log_context(),
+        }
+        current_app.logger.getChild("request").log(
+            logging.ERROR if response.status_code // 100 == 5 else logging.INFO,
```

Comment:
🤔 unsure on this. It feels right, but it will probably trigger an error in sentry for each 5xx request, when we'll already be getting (more useful) error reports from the exceptions themselves.

Reply:
also this log line isn't actually an error itself, it's working correctly, so I might say stick with info.

could be swayed either way

but the sentry thing may need thinking about

Reply:
Would it be better to be a warning?

I'd like it if we didn't have to rely 100% on sentry to spot errors.

Reply:
Warning is fine with me.

I'd rather we only have to use and configure one place for central error reporting.

Reply:
As for "is it an error at all?", I think I'm trying to catch cases where we end up returning a 5xx which weren't because of an exception being raised - e.g. business logic? Would that get surfaced by anything currently?

Anyway, this whole feature is basically a temporary measure pending a full telemetry solution, so perhaps it's not worth me defending..